File: contents/flood_fill/code/python/flood_fill.py
from collections import namedtuple
from queue import Queue
import numpy as np

Point = namedtuple("Point", "x y")

def inbounds(canvas_shape, p):
    return min(p) >= 0 and p.x < canvas_shape[0] and p.y < canvas_shape[1]

def find_neighbors(canvas, p, old_val, new_val):
    # north, south, east, west neighbors
    possible_neighbors = [
        Point(p.x, p.y+1),
        Point(p.x+1, p.y),
        Point(p.x-1, p.y),
        Point(p.x, p.y-1)
    ]

    # exclude the neighbors that go out of bounds and should not be colored
    neighbors = []
    for possible_neighbor in possible_neighbors:
        if inbounds(canvas.shape, possible_neighbor):
            if canvas[possible_neighbor] == old_val:
                neighbors.append(possible_neighbor)
    return neighbors

def stack_fill(canvas, p, old_val, new_val):
    if old_val == new_val:
        return

    stack = [p]

    while stack:
        cur_loc = stack.pop()
        canvas[cur_loc] = new_val
        stack += find_neighbors(canvas, cur_loc, old_val, new_val)

def queue_fill(canvas, p, old_val, new_val):
    if old_val == new_val:
        return

    q = Queue()
    q.put(p)

    canvas[p] = new_val

    while not q.empty():
        cur_loc = q.get()
        neighbors = find_neighbors(canvas, cur_loc, old_val, new_val)

        for neighbor in neighbors:
            canvas[neighbor] = new_val
            q.put(neighbor)

def recursive_fill(canvas, p, old_val, new_val):
    if old_val == new_val:
        return

    canvas[p] = new_val

    neighbors = find_neighbors(canvas, p, old_val, new_val)
    for neighbor in neighbors:
        recursive_fill(canvas, neighbor, old_val, new_val)

def main():
    grid = np.zeros((5, 5))
    grid[2,:] = 1
    
    answer = np.zeros((5, 5))
    answer[:3,] = 1

    c0 = grid.copy()
    c1 = grid.copy()
    c2 = grid.copy()

    start_loc = Point(0, 0)

    recursive_fill(c0, start_loc, 0, 1)
    queue_fill(c1, start_loc, 0, 1)
    stack_fill(c2, start_loc, 0, 1)

    assert (c0 == answer).all()
    assert (c1 == answer).all()
    assert (c2 == answer).all()

    print("Tests Passed")

if __name__ == "__main__":
    main()

File: contents/flood_fill/code/java/flood_fill.java
package code.java;
import java.util.*;
public class flood_fill {
        static class Point{
            int x;
            int y;
            public Point(int x,int y){
                this.x = x;
                this.y = y;
            }
        }
        public static boolean inbound(int[][] grid, Point p){
            int row = grid.length, col = grid[0].length;
            return (p.x < 0 || p.x >= row || p.y < 0 || p.y >= col);
        }
        public static List<Point> find_neighbors(int[][] grid, Point p, int old_val, int new_val){
            //north, south, east, west neighbors
            List<Point> possible_neighbors = new ArrayList<>();
            List<Point> neighbors = new ArrayList<>();
            possible_neighbors.add(new Point(p.x,p.y + 1));
            possible_neighbors.add(new Point(p.x,p.y - 1));
            possible_neighbors.add(new Point(p.x - 1, p.y));
            possible_neighbors.add(new Point(p.x + 1, p.y));
            //exclude the neighbors that go out of bounds and should not be colored
            for (Point possible_neighbor : possible_neighbors){
                if(inbound(grid,possible_neighbor) && grid[possible_neighbor.x][possible_neighbor.y] == old_val){
                    neighbors.add(possible_neighbor);
                }
            }
            return neighbors;
        }
        public static void stack_fill(int[][] grid, Point p, int old_val, int new_val){
            if (old_val == new_val) return;
            Stack<Point> stack = new Stack<>();
            stack.add(p);
            while (!stack.empty()){
                Point cur = stack.pop();
                grid[cur.x][cur.y] = new_val;
                for (Point neighbor : find_neighbors(grid, cur, old_val, new_val)){
                    stack.add(neighbor);
                }
            }
        }
        public static void recursive_fill(int[][] grid, Point p, int old_val, int new_val){
            if (old_val == new_val) return;
            grid[p.x][p.y] = new_val;
            for (Point neighbor : find_neighbors(grid, p, old_val, new_val)){
                recursive_fill(grid, neighbor, old_val, new_val);
            }
        }
        public static void queue_fill(int[][] grid, Point p, int old_val, int new_val){
            if (old_val == new_val) return;
            Queue<Point> queue = new ArrayDeque<>();
            queue.add(p);
            while (!queue.isEmpty()){
                Point cur = queue.remove();
                for (Point neighbor : find_neighbors(grid, cur, old_val, new_val)){
                    grid[neighbor.x][neighbor.y] = new_val;
                    queue.add(neighbor);
                }
            }
        }
        public static boolean isEqual(int[][] grid, int[][] solution){
            for(int i = 0; i < grid.length; i++){
                for(int j = 0; j < grid[0].length; j++) {
                    if(grid[i][j] != solution[i][j]) return false;
                }
            }
            return true;
        }
        public static void main(String[] args) {
            Point start = new Point(0,0);
            int[][] grid = {
                {0, 0, 1, 0, 0},
                {0, 0, 1, 0, 0},
                {0, 0, 1, 0, 0},
                {0, 0, 1, 0, 0},
                {0, 0, 1, 0, 0}};

            int[][] solution = {
                {1, 1, 1, 0, 0},
                {1, 1, 1, 0, 0},
                {1, 1, 1, 0, 0},
                {1, 1, 1, 0, 0},
                {1, 1, 1, 0, 0}};

           queue_fill(grid, start, 0, 1);
           if(isEqual(grid, solution)) System.out.println("test pass");
           stack_fill(grid, start, 0, 1);
           if(isEqual(grid, solution)) System.out.println("test pass");
           recursive_fill(grid, start, 0, 1);
           if(isEqual(grid, solution)) System.out.println("test pass");
        }
}


